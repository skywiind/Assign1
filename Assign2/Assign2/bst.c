/*****************************************************************************************
*                                           *                                            *
*             CS201-001 Assign 2            * This file defines the implementations of   *
*             Wyatt Fairbanks               * the BST class methods.                     *
*             Dr. John Lusth                *                                            *
*             bst.c                         *                                            *
*             10/22/2018                    *                                            *
*                                           *                                            *
*****************************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "bst.h"
#include "queue.h"

struct bst {
	void (*swap)    (TNODE *v1, TNODE *v2);
	int  (*compare) (void *v1, void *v2);
	void (*display) (void *value, FILE *fp);
	void (*free)    (void *value);
	TNODE *root;
	int    size;
	int    debug;
};

//UTILITY FUNCTIONS ////////////////////////////////////////////////////////////////

void swapNodes(TNODE *val1, TNODE *val2) {
	void *temp = getTNODEvalue(val1);
	setTNODEvalue(val1, getTNODEvalue(val2));
	setTNODEvalue(val2, temp);
	return;
}

int isLeaf(TNODE *n) {
	if (!n) {
		return 0;
	}
	if (getTNODEleft(n) == 0 && getTNODEright(n) == 0) {
		return 1;
	}
	return 0;
}

int min(int val1, int val2) {
	if (val1 > val2) {
		return val2;
	}
	if (val2 > val1) {
		return val1;
	}
	return -1;
}

int minDepth(TNODE *node) {
	if (node == 0) {
		return -1;
	}
	if (getTNODEleft(node) && getTNODEright(node)) {
		return min(minDepth(getTNODEleft(node)), minDepth(getTNODEright(node))) + 1;
	}
	return 0;
}

int max(int val1, int val2) {
	if (val1 > val2) {
		return val1;
	}
	if (val2 > val1) {
		return val2;
	}
	return -1;
}

int maxDepth(TNODE *node) {
	if (node == 0) {
		return -1;
	}
	if (isLeaf(node)) {
		return 0;
	}
	if (getTNODEleft(node) && !getTNODEright(node)) {
		return maxDepth(getTNODEleft(node)) + 1;
	}
	if (!getTNODEleft(node) && getTNODEright(node)) {
		return maxDepth(getTNODEright(node)) + 1;
	}
	if (getTNODEleft(node) && getTNODEright(node)) {
		return max(maxDepth(getTNODEleft(node)), maxDepth(getTNODEright(node))) + 1;
	}
	return 0;
}

int isLeftChild(TNODE *n) {
	if (!n) {
		return 0;
	}
	if (getTNODEleft(getTNODEparent(n)) == n) {
		return 1;
	}
	return 0;
}

int isRightChild(TNODE *n) {
	if (!n) {
		return 0;
	}
	if (getTNODEright(getTNODEparent(n)) == n) {
		return 1;
	}
	return 0;
}

void levelOrderDisplay(TNODE *n, FILE *fp) {
	QUEUE *queue = newQUEUE();
	TNODE *temp = n;
	int i = 0;
	int l = 0;
	int c = 1;
	while (l != maxDepth(n) + 2) {
		if (i == 0) {
			fprintf(fp, "%d : ", l);
			l++;
		}
		else if (i == c) {
			fprintf(fp, "\n%d : ", l);
			c *= 2;
			l++;
			i = 0;
		}
		if (temp) {
			if (isLeaf(temp) == 1) {
				fprintf(fp, "=");
			}
			displayTNODE(temp, fp);
			fprintf(fp, "(");
			displayTNODE(getTNODEparent(temp), fp);
			fprintf(fp, ")");
			if (i == 0 && l == 1) {
				fprintf(fp, "X ");
			}
			else if (isLeftChild(temp) == 1) {
				fprintf(fp, "L ");
			}
			else if (isRightChild(temp) == 1) {
				fprintf(fp, "R ");
			}
			enqueue(queue, getTNODEleft(temp));
			enqueue(queue, getTNODEright(temp));
		}
		else {
			enqueue(queue, NULL);
			enqueue(queue, NULL);
		}
		temp = dequeue(queue);
		i++;
	}
	freeQUEUE(queue);
	return;
}

void inOrderDisplay(BST *t, TNODE *n, FILE *fp) {
	if (getTNODEleft(n)) {
		fprintf(fp, "[");
		inOrderDisplay(t, getTNODEleft(n), fp);
		fprintf(fp, "] ");	
	}
	if (n) {
		setTNODEdisplay(n, t->display);
		displayTNODE(n, fp);
	}
	if (getTNODEright(n)) {
		fprintf(fp, " [");
		inOrderDisplay(t, getTNODEright(n), fp);
		fprintf(fp, "]");
	}
}

void preOrderDisplay(BST *t, TNODE *n, FILE *fp) {
	if (n) {
		setTNODEdisplay(n, t->display);
		displayTNODE(n, fp);
	}
	if (getTNODEleft(n)) {
		fprintf(fp, " [");
		preOrderDisplay(t, getTNODEleft(n), fp);
		fprintf(fp, "]");
	}
	if (getTNODEright(n)) {
		fprintf(fp, " [");
		preOrderDisplay(t, getTNODEright(n), fp);
		fprintf(fp, "]");
	}
}

void postOrderDisplay(BST *t, TNODE *n, FILE *fp) {
	if (getTNODEleft(n)) {
		fprintf(fp, "[");
		postOrderDisplay(t, getTNODEleft(n), fp);
		fprintf(fp, "] ");
	}
	if (getTNODEright(n)) {
		fprintf(fp, "[");
		postOrderDisplay(t, getTNODEright(n), fp);
		fprintf(fp, "] ");
	}
	if (n) {
		setTNODEdisplay(n, t->display);
		displayTNODE(n, fp);
	}
}

void postOrderFree(BST *t, TNODE *n) {
	if (getTNODEleft(n)) {
		postOrderFree(t, getTNODEleft(n));
	}
	if (getTNODEright(n)) {
		postOrderFree(t, getTNODEright(n));
	}
	if (n) {
		setTNODEfree(n, t->free);
		freeTNODE(n);
	}
}

//END UTILITY FUNCS ////////////////////////////////////////////////////////////////

BST *newBST(int(*c) (void *, void*)) {
	BST *tree     = malloc(sizeof(BST));
	tree->root    = 0;
	tree->size    = 0;
	tree->debug   = 0;
	tree->swap    = swapNodes;
	tree->compare = c;
	tree->display = 0;
	tree->free    = 0;
	return tree;
}

void setBSTdisplay(BST *t, void (*d)(void *, FILE *)) {
	t->display = d;
	return;
}

void setBSTswapper(BST *t, void (*s)(TNODE *, TNODE *)) {
	t->swap = s;
	return;
}

void setBSTfree(BST *t, void(*f)(void *)) {
	t->free = f;
	return;
}

TNODE *getBSTroot(BST *t) {
	return t->root;
}

void setBSTroot(BST *t, TNODE *replacement) {
	t->root = replacement;
	return;
}

void setBSTsize(BST *t, int s) {
	t->size = s;
	return;
}

TNODE *insertBST(BST *t, void *value) {
	TNODE *node    = t->root;
	TNODE *newNode = newTNODE(value, 0, 0, 0);
	if (t->display) {
		setTNODEdisplay(newNode, t->display);
	}
	if (node == 0) {
		setTNODEparent(newNode, newNode);
		t->root = newNode;
		setBSTsize(t, 1);
		return newNode;
	}
	while (node != 0) {
		int c = t->compare(getTNODEvalue(node), value);
		if (c > 0 && getTNODEleft(node) == 0) {
			setTNODEleft(node, newNode);
			setTNODEparent(newNode, node);
			setBSTsize(t, t->size + 1);
			return newNode;
		}
		if (c < 0 && getTNODEright(node) == 0) {
			setTNODEright(node, newNode);
			setTNODEparent(newNode, node);
			setBSTsize(t, t->size + 1);
			return newNode;
		}
		if (c > 0) {
			node = getTNODEleft(node);
		}
		if (c < 0) {
			node = getTNODEright(node);
		}
		if (c == 0) {
			return NULL;
		}
	}
	return NULL;
}

void *findBST(BST *t, void *value) {
	TNODE *node = t->root;
	if (node == 0) {
		return NULL;
	}
	while (node != 0) {
		int c = t->compare(getTNODEvalue(node), value);
		if (c == 0) {
			return getTNODEvalue(node);
		}
		else {
			if (c < 0) {
				node = getTNODEleft(node);
			}
			if (c > 0) {
				node = getTNODEright(node);
			}
		}
	}
	return NULL;
}

TNODE *locateBST(BST *t, void *value) {
	TNODE *node = t->root;
	if (node == 0) {
		return NULL;
	}
	while (node != 0) {
		int c = t->compare(getTNODEvalue(node), value);
		if (c == 0) {
			return node;
		}
		else {
			if (c < 0) {
				node = getTNODEleft(node);
			}
			if (c > 0) {
				node = getTNODEright(node);
			}
		}
	}
	return NULL;
}

int deleteBST(BST *t, void *key) {
	TNODE *target = locateBST(t, key);
	if (target != NULL) {
		return -1;
	}
	target = swapToLeafBST(t, target);
	pruneLeafBST(t, target);
	if (t->free) {
		t->free(getTNODEvalue(target));
	}
	free(target);
	setBSTsize(t, t->size - 1);
	if (t->size == 0) {
		t->root = NULL;
	}
	return 0;
}

TNODE *swapToLeafBST(BST *t, TNODE *node) {
	if (isLeaf(node) == 1) {
		return node;
	}
	if (getTNODEleft(node)) {
		t->swap(node, getTNODEleft(node));
		node = getTNODEleft(node);
		while (getTNODEright(node)) {
			t->swap(node, getTNODEright(node));
			node = getTNODEright(node);
		}
		return swapToLeafBST(t, node);
	}
	else if (getTNODEright(node)) {
		t->swap(node, getTNODEright(node));
		node = getTNODEright(node);
		while (getTNODEleft(node)) {
			t->swap(node, getTNODEleft(node));
			node = getTNODEleft(node);
		}
		return swapToLeafBST(t, node);
	}
	return node;
}

void pruneLeafBST(BST *t, TNODE *leaf) {
	TNODE *parent = getTNODEparent(leaf);
	if (getBSTroot(t) == leaf) {
		return;
	}	
	if (leaf == getTNODEleft(parent)) {
		setTNODEleft(parent, 0);
		return;
	}
	if (leaf == getTNODEright(parent)) {
		setTNODEright(parent, 0);
		return;
	}
}

int sizeBST(BST *t) {
	return t->size;
}

void statisticsBST(BST *t, FILE *fp) {
	int size = sizeBST(t);
	if (size == 0) {
		fprintf(fp, "Nodes: 0\nMinimum depth: -1\nMaximum depth: -1\n");
		return;
	}
	fprintf(fp, "Nodes: %d\nMinimum depth: %d\nMaximum depth: %d\n", size, minDepth(t->root), maxDepth(t->root));
	return;
}

void displayBST(BST *t, FILE *fp) {
	if (getBSTroot(t) == 0) {
		fprintf(fp, "0\n");
		return;
	}

	if (t->debug > 0 && t->root == NULL) {
		fprintf(fp, "[]\n");
		return;
	}
	else if (t->debug == 1) {
		fprintf(fp, "[");
		inOrderDisplay(t, t->root, fp);
		fprintf(fp, "]\n");
	}
	else if (t->debug == 2) {
		fprintf(fp, "[");
		preOrderDisplay(t, t->root, fp);
		fprintf(fp, "]\n");
	}
	else if (t->debug == 3) {
		fprintf(fp, "[");
		postOrderDisplay(t, t->root, fp);
		fprintf(fp, "]\n");
	}
	else {
		levelOrderDisplay(getBSTroot(t), fp);
	}
	return;
}

int debugBST(BST *t, int level) {
	int temp = t->debug;
	t->debug = level;
	return temp;
}

void freeBST(BST *t) {
	postOrderFree(t, t->root);
	free(t);
}