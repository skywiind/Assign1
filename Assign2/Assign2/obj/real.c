#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL{
    double value;
};

REAL* newREAL(double x){
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
}

double getREAL(REAL *v){
    return v->value;
}

double setREAL(REAL *v,double x){
    double old = v->value;
    v->value = x;
    return old;
}

void displayREAL(void *v,FILE *fp){
    fprintf(fp,"%lf",getREAL((REAL*) v));
	return;
}

double compareREAL(void *v,void *w){
    return getREAL(v) - getREAL(w);
}

void freeREAL(void *v){
    free((REAL *) v);
	return;
}
