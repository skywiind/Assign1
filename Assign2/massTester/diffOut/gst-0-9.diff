3c3
< Duplicates: 2087
---
> Duplicates: 1517
22c22
< Duplicates: 1202
---
> Duplicates: 31
