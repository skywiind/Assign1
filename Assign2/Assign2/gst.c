/*****************************************************************************************
*                                           *                                            *
*             CS201-001 Assign 2            * This file defines the implementations of   *
*             Wyatt Fairbanks               * the GST class methods.                     *
*             Dr. John Lusth                *                                            *
*             gst.c                         *                                            *
*             10/22/2018                    *                                            *
*                                           *                                            *
*****************************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include "bst.h"
#include "gst.h"

struct gst {
	BST   *tree;
	void (*display) (void *value, FILE *fp);
	void (*free) (void *value);
	void (*swap) (TNODE *v1, TNODE *v2);
	int  (*compare) (void *v1, void *v2);
	int d;
};

typedef struct gnode GNODE;

struct gnode {
	void  *v;
	int    f;
	void (*display) (void *value, FILE *fp);
	void (*free) (void *value);
	int  (*compare) (void *v1, void *v2);
};

//UTILITY FUNCTIONS ////////////////////////////////////////////////////////////////

void displayGNODE(void *v, FILE *fp) {
	GNODE *temp = v;
	if (temp == NULL) {
		return;
	}
	temp->display(temp->v, fp);
	GNODE *test = temp->v;
	if (test->f > 1) {
		fprintf(fp, "<%d>", test->f);
	}
	/*
	if (temp->f > 1) {
		fprintf(fp, "<%d>", temp->f);
	}
	*/
	return;
}

int compareGNODE(void *v, void *w) {
	GNODE *a = v;
	GNODE *b = w;
	return b->compare(a->v, b->v);
}

void freeGNODE(void *n) {
	GNODE *temp = n;
	temp->free(temp->v);
	free(temp);
	return;
}

void swapper(TNODE *a, TNODE *b) {
	void *x = getTNODEvalue(a);
	void *y = getTNODEvalue(b);
	setTNODEvalue(a, y);
	setTNODEvalue(b, x);
	return;
}

//END UTILITY FUNCS ////////////////////////////////////////////////////////////////

GNODE *newGNODE(void *value) {
	GNODE *newNode = malloc(sizeof(GNODE));
	newNode->v = value;
	newNode->f = 1;
	return newNode;
}

GST *newGST(int(*c)(void *, void *)) {
	GST *newTree     = malloc(sizeof(GST));
	newTree->tree    = newBST(compareGNODE);
	setBSTdisplay(newTree->tree, displayGNODE);
	setBSTfree(newTree->tree, freeGNODE);
	newTree->display = 0;
	newTree->free    = 0;
	newTree->compare = c;
	newTree->swap    = swapper;
	newTree->d       = 0;
	return newTree;
}

void setGSTdisplay(GST *t, void(*d)(void *v, FILE *fp)) {
	t->display = d;
	return;
}

void setGSTswapper(GST *t, void(*s)(TNODE *, TNODE *)) {
	t->swap = s;
	setBSTswapper(t->tree, s);
	return;
}

void setGSTfree(GST *t, void(*f)(void *)) {
	t->free = f;
	return;
}

TNODE *getGSTroot(GST *t) {
	return getBSTroot(t->tree);
}

void setGSTroot(GST *t, TNODE *replacement) {
	setBSTroot(t->tree, replacement);
	return;
}
void setGSTsize(GST *t, int s) {
	setBSTsize(t->tree, s);
	return;
}

TNODE *insertGST(GST *t, void *value) {
	GNODE *newNode   = newGNODE(value);
	newNode->display = t->display;
	newNode->compare = t->compare;
	newNode->free    = t->free;

	if (findBST(t->tree, newNode)) {
		TNODE *temp = locateGST(t, newNode);
		GNODE *unwrap = unwrapGST(temp);
		unwrap->f++;
		t->d++;
		freeGNODE(newNode);
		return temp;
	}

	return insertBST(t->tree, newNode);
}

void *findGST(GST *t, void *key) {
	return unwrapGST(findBST(t->tree, key));
}

TNODE *locateGST(GST *t, void *key) {
	return locateBST(t->tree, key);
}

int deleteGST(GST *t, void *key) {
	GNODE *temp = unwrapGST(locateGST(t, key));
	if (temp == 0) {
		return -1;
	}
	if (temp->f > 1) {
		temp->f--;
		return temp->f;
	}
	else {
		return deleteBST(t->tree, key);
	}
	return -1;
}

TNODE *swapToLeafGST(GST *t, TNODE *node) {
	return swapToLeafBST(t->tree, node);
}

void pruneLeafGST(GST *t, TNODE *leaf) {
	return pruneLeafBST(t->tree, leaf);
}

int sizeGST(GST *t) {
	return sizeBST(t->tree);
}

void statisticsGST(GST *t, FILE *fp) {
	fprintf(fp, "Duplicates: %d\n", t->d);
	statisticsBST(t->tree, fp);
	return;
}

void displayGST(GST *t, FILE *fp) {
	displayBST(t->tree, fp);
	return;
}

void freeGST(GST *t) {
	freeBST(t->tree);
	free(t);
	return;
}

int debugGST(GST *t, int level) {
	return debugBST(t->tree, level);
}

void *unwrapGST(TNODE *node) {
	if (!node) {
		return NULL;
	}
	GNODE *temp = getTNODEvalue(node);
	return temp->v;
}

int freqGST(GST *g, void *key) {
	TNODE *temp = locateGST(g, key);
	if (temp) {
		GNODE *unwrap = getTNODEvalue(temp);
		return unwrap->f;
	}
	return 0;
}

int duplicatesGST(GST *g) {
	return g->d;
}