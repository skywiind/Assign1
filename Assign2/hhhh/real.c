#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL{
    double value;
};

REAL* newREAL(double x){
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
}

double getREAL(REAL *v){
    return v->value;
}

double setREAL(REAL *v,double x){
    double old = v->value;
    v->value = x;
    return old;
}

void displayREAL(void *v,FILE *fp){
    fprintf(fp,"%lf",getREAL((REAL*) v));
	return;
}

int compareREAL(void *v,void *w){
    double r = getREAL(v) - getREAL(w);
    if (r < 0.0) {
	    return -1;
    } else if (r > 0.0) {
	    return 1;
    } else {
	    return 0;
    }
}

void freeREAL(void *v){
    free((REAL *) v);
	return;
}
