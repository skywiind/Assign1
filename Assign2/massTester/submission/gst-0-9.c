#include <stdio.h>
#include <stdlib.h>
#include "bst.h"
#include "gst.h"
#include "integer.h"
#include "real.h"
#include "string.h"

void srandom(unsigned int);
long int random(void);

int
main(void)
    {
    srandom(21);
    printf("INTEGER test of GST, mixed insertions and deletions\n");
    int i;
    GST *p = newGST(compareINTEGER);
    setGSTdisplay(p,displayINTEGER);
    setGSTfree(p,freeINTEGER);
    for (i = 0; i < 20000; ++i)
        {
        int j = random() % 3333;
        INTEGER *a = newINTEGER(j);
        if (random() % 10 < 6)
            {
            if (freqGST(p,a) > 1)
                deleteGST(p,a);
            else
                {
                INTEGER *x = findGST(p,a);
                setGSTfree(p,0);
                deleteGST(p,a);
                setGSTfree(p,freeINTEGER);
                freeINTEGER(x);
                }
            freeINTEGER(a);
            }
        else
             insertGST(p,a);
        }
    if (sizeGST(p) < 200)
        {
        debugGST(p,0);
        printf("GST:\n");
        displayGST(p,stdout);
        debugGST(p,1);
        printf("debug (in-order): ");
        displayGST(p,stdout);
        printf("\n");
        debugGST(p,2);
        printf("debug (pre-order): ");
        displayGST(p,stdout);
        printf("\n");
        debugGST(p,3);
        printf("debug (post-order): ");
        displayGST(p,stdout);
        printf("\n");
        }
    printf("size is %d\n",sizeGST(p));
    statisticsGST(p,stdout);
    while (sizeGST(p) > 200)
        {
        for (i = 0; i < 10000; ++i)
            {
            int j = random() % 5000;
            INTEGER *a = newINTEGER(j);
            if (freqGST(p,a) > 1)
                deleteGST(p,a);
            else
                {
                INTEGER *x = findGST(p,a);
                setGSTfree(p,0);
                deleteGST(p,a);
                setGSTfree(p,freeINTEGER);
                freeINTEGER(x);
                }
            freeINTEGER(a);
            }
        }
    if (sizeGST(p) < 200)
        {
        debugGST(p,0);
        printf("GST:\n");
        displayGST(p,stdout);
        debugGST(p,1);
        printf("debug: ");
        displayGST(p,stdout);
        printf("\n");
        debugGST(p,2);
        printf("debug: ");
        displayGST(p,stdout);
        printf("\n");
        debugGST(p,3);
        printf("debug: ");
        displayGST(p,stdout);
        printf("\n");
        }
    printf("size is %d\n",sizeGST(p));
    statisticsGST(p,stdout);
    setGSTfree(p,freeINTEGER);
    freeGST(p);
    return 0;
    }
