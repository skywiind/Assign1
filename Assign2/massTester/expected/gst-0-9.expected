INTEGER test of GST, mixed insertions and deletions
size is 1839
Duplicates: 1517
Nodes: 1839
Minimum depth: 3
Maximum depth: 23
GST:
0: 2998<2>(2998<2>)X
1: 1140(2998<2>)L 3079<2>(2998<2>)R
2: 918(1140)L 2108(1140)R =3066<4>(3079<2>)L =3125(3079<2>)R
3: 331<2>(918)L =1031(918)R 1688(2108)L 2441<2>(2108)R
4: 310<5>(331<2>)L 440<2>(331<2>)R 1597(1688)L 1722<2>(1688)R 2774(2441<2>)R
5: 97(310<5>)L 792<2>(440<2>)R 1388(1597)L 1682(1597)R =2041<4>(1722<2>)R 2652(2774)L 2867<3>(2774)R
6: =74<2>(97)L 187(97)R 781(792<2>)L 855(792<2>)R 1483(1388)R =1676(1682)L 2587<2>(2652)L 2685<2>(2652)R 2863(2867<3>)L =2926(2867<3>)R
7: 231(187)R =683(781)L 832(855)L 1530<2>(1483)R 2484<2>(2587<2>)L 2600<2>(2587<2>)R =2677<3>(2685<2>)L =2752(2685<2>)R =2822(2863)L
8: 207(231)L =814<2>(832)L 1541(1530<2>)R =2525(2484<2>)R =2616(2600<2>)R
9: =206<2>(207)L =1589<3>(1541)R
debug: [[[[[[[74<2>] 97 [187 [[[206<2>] 207] 231]]] 310<5>] 331<2> [440<2> [[[683] 781] 792<2> [[[814<2>] 832] 855]]]] 918 [1031]] 1140 [[[[1388 [1483 [1530<2> [1541 [1589<3>]]]]] 1597 [[1676] 1682]] 1688 [1722<2> [2041<4>]]] 2108 [2441<2> [[[[2484<2> [2525]] 2587<2> [2600<2> [2616]]] 2652 [[2677<3>] 2685<2> [2752]]] 2774 [[[2822] 2863] 2867<3> [2926]]]]]] 2998<2> [[3066<4>] 3079<2> [3125]]]
debug: [2998<2> [1140 [918 [331<2> [310<5> [97 [74<2>] [187 [231 [207 [206<2>]]]]]] [440<2> [792<2> [781 [683]] [855 [832 [814<2>]]]]]] [1031]] [2108 [1688 [1597 [1388 [1483 [1530<2> [1541 [1589<3>]]]]] [1682 [1676]]] [1722<2> [2041<4>]]] [2441<2> [2774 [2652 [2587<2> [2484<2> [2525]] [2600<2> [2616]]] [2685<2> [2677<3>] [2752]]] [2867<3> [2863 [2822]] [2926]]]]]] [3079<2> [3066<4>] [3125]]]
debug: [[[[[[[74<2>] [[[[206<2>] 207] 231] 187] 97] 310<5>] [[[[683] 781] [[[814<2>] 832] 855] 792<2>] 440<2>] 331<2>] [1031] 918] [[[[[[[[1589<3>] 1541] 1530<2>] 1483] 1388] [[1676] 1682] 1597] [[2041<4>] 1722<2>] 1688] [[[[[[2525] 2484<2>] [[2616] 2600<2>] 2587<2>] [[2677<3>] [2752] 2685<2>] 2652] [[[2822] 2863] [2926] 2867<3>] 2774] 2441<2>] 2108] 1140] [[3066<4>] [3125] 3079<2>] 2998<2>]
size is 49
Duplicates: 31
Nodes: 49
Minimum depth: 2
Maximum depth: 9
